BlueprintUICommonControls/Sources/AttributedLabel.swift

```diff
@@ -4,34 +4,43 @@ import UIKit
 public struct AttributedLabel: Element {
 
     public var attributedText: NSAttributedString
-    public var numberOfLines: Int = 0
+    public var numberOfLines: Int
     /// The scale to which pixel measurements will be rounded. Defaults to `UIScreen.main.scale`.
     public var roundingScale: CGFloat = UIScreen.main.scale
 
-    public init(attributedText: NSAttributedString) {
+    public init(attributedText: NSAttributedString, numberOfLines: Int = 0) {
         self.attributedText = attributedText
+        self.numberOfLines = numberOfLines
     }
+    
+    /**
+     Keep around a label to use for measurement and sizing.
+     
+     We would usually do this using NSString or NSAttributedString's boundingRect family of methods,
+     but these do not let you specify a `numberOfLines` parameter, which is critical to correct sizing.
+     
+     As such, we will allocate this label once and then use it to measure by setting its text and attributes.
+     */
+    static let measurementLabel = UILabel()
 
     public var content: ElementContent {
-        struct Measurer: Measurable {
-
-            var attributedText: NSAttributedString
-            var roundingScale: CGFloat
-
-            func measure(in constraint: SizeConstraint) -> CGSize {
-                var size = attributedText.boundingRect(
-                    with: constraint.maximum,
-                    options: [.usesLineFragmentOrigin],
-                    context: nil)
-                    .size
-                size.width = size.width.rounded(.up, by: roundingScale)
-                size.height = size.height.rounded(.up, by: roundingScale)
-
-                return size
-            }
+        
+        return ElementContent { constraint in
+            let label = AttributedLabel.measurementLabel
+            let description = self.backingViewDescription(bounds: CGRect(origin: .zero, size: constraint.maximum), subtreeExtent: nil)!
+            
+            description.apply(to: label)
+            label.attributedText = self.attributedText
+            
+            var size = label.sizeThatFits(constraint.maximum)
+            
+            size.width = size.width.rounded(.up, by: self.roundingScale)
+            size.height = size.height.rounded(.up, by: self.roundingScale)
```

Comment:
Confirmed that it rounds to the screen scale:

```
(lldb) po size
▿ (153.5, 38.5)
  - width : 153.5
  - height : 38.5
```

Reply:
Fwiw, not rounding to full px values does make the snapshots look a bit odd (new on right):

<img width="1792" alt="Screen Shot 2020-02-22 at 9 31 11 PM" src="https://user-images.githubusercontent.com/327847/75104184-a8a6c080-55ba-11ea-8f0e-5ad6cde40f59.png">

Reply:
I remembered why we needed this, now: Because the label has `roundingScale`, we need to do this round up in order to conform to it. Otherwise the rounding returned is always the current device's scale, which is only right if you don't change `roundingScale`.